roadmap/implementers-guide/src/node/utility/availability-store.md

```diff
@@ -65,104 +59,127 @@ digraph {
 
-On `OverseerSignal::BlockFinalized(_)` events:
+```rust
+struct CandidateMeta {
+  state: State,
+  data_available: bool,
+  chunks_stored: Bitfield,
+}
 
-- Handle all pruning based on the newly-finalized block.
+enum State {
+  /// Candidate data was first observed at the given time but is not available in any block.
+  Unavailable(Timestamp),
+  /// The candidate was first observed at the given time and was included in the given list of unfinalized blocks, which may be
+  /// empty. The timestamp here is not used for pruning.
+  Unfinalized(Timestamp, Vec<(BlockNumber, BlockHash)>),
+  /// Candidate data has appeared in a finalized block and did so at the given time.
+  Finalized(Timestamp)
+}
+```
 
-On `QueryPoV` message:
+We maintain the invariant that if a candidate has a meta entry, its available data exists on disk if `data_available` is true. All chunks mentioned in the meta entry are available.
 
-- Return the PoV block, if any, for that candidate hash.
+Additionally, there is exactly one `prune_by_time` entry which holds the candidate hash unless the state if `Unfinalized`. There may be zero, one, or many "appears_in_unfinalized" keys with the given candidate, and this will correspond to the `state` of the meta entry.
 
-On `QueryChunk` message:
+## Protocol
 
-- Determine if we have the chunk indicated by the parameters and return it and its inclusion proof via the response channel if so.
+Input: [`AvailabilityStoreMessage`][ASM]
+
+Output:
+- [`RuntimeApiMessage`][RAM]
 
-On `StoreChunk` message:
 
-- Store the chunk along with its inclusion proof under the candidate hash and validator index.
+## Functionality
 
-On `StorePoV` message:
+For each head in the `activated` list:
+	- Note any new candidates backed in the block. Update the `CandidateMeta` for each. If the `CandidateMeta` does not exist, create it as `Unavailable` with the current timestamp. Register a `"prune_by_time"` entry based on the current timestamp + 1 hour.
+  - Note any new candidate included in the block. Update the `CandidateMeta` for each, performing a transition from `Unavailable` to `Unfinalized` if necessary. That includes removing the `"prune_by_time"` entry. Add the block hash and number to the state, if unfinalized. If the head is final already, just set the state to `State::Finalized`. Set pruning info accordingly.
+  - The `CandidateEvent` runtime API can be used for this purpose.
+  - TODO: load all ancestors of the head back to the finalized block so we don't miss anything if import notifications are missed. If a `StoreChunk` message is received for a candidate which has no entry, then we will prematurely lose the data.
 
-- Store the block, if the validator index is provided, store the respective chunk as well.
+On `OverseerSignal::BlockFinalized(finalized)` events:
+  - for each key in `iter_by_prefix(("appears_in_unfinalized", last_finalized))`
+    - Stop if the key is beyond `("appears_in_unfinalized, finalized)`
```

Comment:
that sounds like the block number should be in big endian so that lexicographical order yields values by ascension

Reply:
Yes

Reply:
Timestamps as well